File: examples/pub2.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * Copyright (c) 2019      IBM Corporation.  All rights reserved.
 * Copyright (c) 2021-2025 Nanook Consulting  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#define _GNU_SOURCE
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <libgen.h>

#include <pmix.h>

#include "examples.h"

#define NITER 3

int main(int argc, char **argv)
{
    pmix_proc_t myproc;
    int rc, n;
    pmix_value_t *val = NULL;
    pmix_proc_t proc;
    uint32_t nprocs;
    pmix_info_t info[2], lkinfo[2];
    pmix_pdata_t pdata;
    pmix_persistence_t firstread = PMIX_PERSIST_FIRST_READ;
    int timeout=10;
    char *tmp;
    int iters;

    if (1 < argc) {
        iters = strtol(argv[1], NULL, 10);
    } else {
        iters = NITER;
    }

    /* init us */
    if (PMIX_SUCCESS != (rc = PMIx_Init(&myproc, NULL, 0))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Init failed: %d\n", myproc.nspace, myproc.rank,
                rc);
        exit(0);
    }

    /* get our job size */
    PMIX_PROC_CONSTRUCT(&proc);
    PMIX_LOAD_PROCID(&proc, myproc.nspace, PMIX_RANK_WILDCARD);
    if (PMIX_SUCCESS != (rc = PMIx_Get(&proc, PMIX_JOB_SIZE, NULL, 0, &val))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Get job size failed: %d\n", myproc.nspace,
                myproc.rank, rc);
        goto done;
    }
    nprocs = val->data.uint32;
    PMIX_VALUE_RELEASE(val);
    // require at least 2 procs
    if (nprocs != 2) {
        if (0 == myproc.rank) {
            fprintf(stderr, "%s requires 2 processes\n", basename(argv[0]));
        }
        exit(1);
    }

    PMIX_INFO_LOAD(&info[1], PMIX_PERSISTENCE, &firstread, PMIX_PERSIST);
    PMIX_INFO_LOAD(&lkinfo[0], PMIX_WAIT, NULL, PMIX_BOOL);
    PMIX_INFO_LOAD(&lkinfo[1], PMIX_TIMEOUT, &timeout, PMIX_INT);
    for (n=0; n < iters; n++) {
        if (0 == myproc.rank) {
            /* publish something */
            if (0 > asprintf(&tmp, "FOOBAR:%s.%u:%d", myproc.nspace, myproc.rank, n)) {
                goto done;
            }
            PMIX_INFO_LOAD(&info[0], tmp, &n, PMIX_INT);
            free(tmp);
            if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 2))) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                        myproc.rank, rc);
                goto done;
            }
            PMIX_INFO_DESTRUCT(&info[0]);

            /* lookup other rank's value */
            PMIX_PDATA_CONSTRUCT(&pdata);
            if (0 > asprintf(&tmp, "BAZ:%s.%u:%d", myproc.nspace, 1, n)) {
                goto done;
            }
            PMIX_LOAD_KEY(pdata.key, tmp);
            free(tmp);
            // lookup value
            if (PMIX_SUCCESS != (rc = PMIx_Lookup(&pdata, 1, lkinfo, 2))) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup failed: %d\n", myproc.nspace,
                        myproc.rank, rc);
                goto done;
            }
            /* check the return for value and source */
            if (0 != strncmp(myproc.nspace, pdata.proc.nspace, PMIX_MAX_NSLEN)) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s\n",
                        myproc.nspace, myproc.rank, pdata.proc.nspace);
                goto done;
            }
            if (1 != pdata.proc.rank) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d\n",
                        myproc.nspace, myproc.rank, pdata.proc.rank);
                goto done;
            }
            if (PMIX_INT != pdata.value.type) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong type: %d\n",
                        myproc.nspace, myproc.rank, pdata.value.type);
                goto done;
            }
            if (n != pdata.value.data.integer) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong value: %d\n",
                        myproc.nspace, myproc.rank, (int) pdata.value.data.uint8);
                goto done;
            }
            PMIX_PDATA_DESTRUCT(&pdata);
            fprintf(stderr, "PUBLISH-LOOKUP SUCCEEDED: %d\n", n);
        } else {
            /* lookup other rank's value */
            PMIX_PDATA_CONSTRUCT(&pdata);
            if (0 > asprintf(&tmp, "FOOBAR:%s.%u:%d", myproc.nspace, 0, n)) {
                goto done;
            }
            PMIX_LOAD_KEY(pdata.key, tmp);
            free(tmp);
            // check value
            if (PMIX_SUCCESS != (rc = PMIx_Lookup(&pdata, 1, lkinfo, 2))) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup failed: %d\n", myproc.nspace,
                        myproc.rank, rc);
                goto done;
            }
            /* check the return for value and source */
            if (0 != strncmp(myproc.nspace, pdata.proc.nspace, PMIX_MAX_NSLEN)) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s\n",
                        myproc.nspace, myproc.rank, pdata.proc.nspace);
                goto done;
            }
            if (0 != pdata.proc.rank) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d\n",
                        myproc.nspace, myproc.rank, pdata.proc.rank);
                goto done;
            }
            if (PMIX_INT != pdata.value.type) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong type: %d\n",
                        myproc.nspace, myproc.rank, pdata.value.type);
                goto done;
            }
            if (n != pdata.value.data.integer) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Lookup returned wrong value: %d\n",
                        myproc.nspace, myproc.rank, (int) pdata.value.data.uint8);
                goto done;
            }
            PMIX_PDATA_DESTRUCT(&pdata);
            /* publish something */
            if (0 > asprintf(&tmp, "BAZ:%s.%u:%d", myproc.nspace, myproc.rank, n)) {
                goto done;
            }
            PMIX_INFO_LOAD(&info[0], tmp, &n, PMIX_INT);
            free(tmp);
            if (PMIX_SUCCESS != (rc = PMIx_Publish(info, 2))) {
                fprintf(stderr, "Client ns %s rank %d: PMIx_Publish failed: %d\n", myproc.nspace,
                        myproc.rank, rc);
                goto done;
            }
            PMIX_INFO_DESTRUCT(&info[0]);
        }
    }

done:
    /* finalize us */
    if (PMIX_SUCCESS != (rc = PMIx_Finalize(NULL, 0))) {
        fprintf(stderr, "Client ns %s rank %d:PMIx_Finalize failed: %d\n", myproc.nspace,
                myproc.rank, rc);
    } else {
        fprintf(stderr, "Client ns %s rank %d:PMIx_Finalize successfully completed\n",
                myproc.nspace, myproc.rank);
    }
    fflush(stderr);
    return (0);
}

File: examples/python/pub2.py
#!/usr/bin/env python3
#
# Copyright (c) 2026      Nanook Consulting  All rights reserved.
# $COPYRIGHT$
#
# Additional copyrights may follow
#
# $HEADER$
#
# Python port of examples/pub2.c
#
# Two ranks ping-pong through the publish/lookup service: each iteration
# rank 0 publishes FOOBAR:<n> and looks up rank 1's BAZ:<n>, while rank 1
# does the reverse. The published data is marked PMIX_PERSIST_FIRST_READ
# so it is consumed by the matching lookup, and the lookups carry
# PMIX_WAIT so they block until the peer publishes.
#
# Takes an optional iteration count; requires exactly 2 processes.

import os
import sys

from examples import *

NITER = 3


def main():
    if 1 < len(sys.argv):
        iters = int(sys.argv[1])
    else:
        iters = NITER

    client = PMIxClient()

    # init us
    rc, myproc = client.init(None)
    if PMIX_SUCCESS != rc:
        eprint("Client: PMIx_Init failed: %d" % rc)
        sys.exit(0)

    # get our job size
    proc = {'nspace': myproc['nspace'], 'rank': PMIX_RANK_WILDCARD}
    rc, val = client.get(proc, PMIX_JOB_SIZE, None)
    if PMIX_SUCCESS != rc:
        eprint("Client ns %s rank %d: PMIx_Get job size failed: %d"
               % (myproc['nspace'], myproc['rank'], rc))
        return done(client, myproc)
    nprocs = val['value']
    # require at least 2 procs
    if 2 != nprocs:
        if 0 == myproc['rank']:
            eprint("%s requires 2 processes"
                   % os.path.basename(sys.argv[0]))
        sys.exit(1)

    # the persistence directive rides along with every publish
    persist = {'key': PMIX_PERSISTENCE, 'value': PMIX_PERSIST_FIRST_READ,
               'val_type': PMIX_PERSIST}
    lkinfo = [{'key': PMIX_WAIT, 'value': True, 'val_type': PMIX_BOOL},
              {'key': PMIX_TIMEOUT, 'value': 10, 'val_type': PMIX_INT}]

    for n in range(iters):
        if 0 == myproc['rank']:
            # publish something
            key = "FOOBAR:%s.%u:%d" % (myproc['nspace'], myproc['rank'], n)
            info = [{'key': key, 'value': n, 'val_type': PMIX_INT}, persist]
            rc = client.publish(info)
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Publish failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)

            # lookup other rank's value
            key = "BAZ:%s.%u:%d" % (myproc['nspace'], 1, n)
            rc, pdata = client.lookup([{'key': key}], lkinfo)
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Lookup failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)
            if not check(client, myproc, pdata[0], 1, n):
                return done(client, myproc)
            eprint("PUBLISH-LOOKUP SUCCEEDED: %d" % n)
        else:
            # lookup other rank's value
            key = "FOOBAR:%s.%u:%d" % (myproc['nspace'], 0, n)
            rc, pdata = client.lookup([{'key': key}], lkinfo)
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Lookup failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)
            if not check(client, myproc, pdata[0], 0, n):
                return done(client, myproc)

            # publish something
            key = "BAZ:%s.%u:%d" % (myproc['nspace'], myproc['rank'], n)
            info = [{'key': key, 'value': n, 'val_type': PMIX_INT}, persist]
            rc = client.publish(info)
            if PMIX_SUCCESS != rc:
                eprint("Client ns %s rank %d: PMIx_Publish failed: %d"
                       % (myproc['nspace'], myproc['rank'], rc))
                return done(client, myproc)

    return done(client, myproc)


def check(client, myproc, pdata, rank, n):
    """Check a lookup return for value and source."""
    if myproc['nspace'] != pdata['proc']['nspace']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong nspace: %s"
               % (myproc['nspace'], myproc['rank'], pdata['proc']['nspace']))
        return False
    if rank != pdata['proc']['rank']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong rank: %d"
               % (myproc['nspace'], myproc['rank'], pdata['proc']['rank']))
        return False
    if PMIX_INT != pdata['val_type']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong type: %d"
               % (myproc['nspace'], myproc['rank'], pdata['val_type']))
        return False
    if n != pdata['value']:
        eprint("Client ns %s rank %d: PMIx_Lookup returned wrong value: %d"
               % (myproc['nspace'], myproc['rank'], pdata['value']))
        return False
    return True


def done(client, myproc):
    # finalize us
    rc = client.finalize(None)
    if PMIX_SUCCESS != rc:
        eprint("Client ns %s rank %d:PMIx_Finalize failed: %d"
               % (myproc['nspace'], myproc['rank'], rc))
    else:
        eprint("Client ns %s rank %d:PMIx_Finalize successfully completed"
               % (myproc['nspace'], myproc['rank']))
    return 0


if __name__ == '__main__':
    sys.exit(main())
